File: analyze_ast.py
import sys
import json


# 함수 개수 추출 (희진)
def parse_func_cnt(data):
    func_cnt = 0
    
    for i in data["ext"]:
        if i["_nodetype"]=="FuncDef":
            func_cnt+=1
    return func_cnt


# 함수 리턴 타입 추출 (하영)
def parse_func_ret(data):
    func_ret = list()
    
    # write your code
    
    return func_ret


# 함수 이름 추출 (소정)
def parse_func_name(data):
    func_name = list()

    for i in data["ext"]:
        if i["_nodetype"]=="FuncDef":
            func_name.append(i['decl']['name'])
    
    return func_name


# 파라미터 타입과 변수명 추출 (다인)
def parse_param(data):
    param_type = list()
    param_name = list()
    
    # write your code
    
    return param_type, param_name


# if 조건 개수 추출 (송현)
def parse_if_cnt(data):
    if_cnt = 0
    
    # write your code
    
    return if_cnt


def print_analysis(func_cnt, func_ret, func_name, param_type, param_name, if_cnt):

    for i in range(func_cnt):
        print("Function", i+1)
        print("- Name:", func_name[i])
        print("- Return Type:")
        print("- Parameter Type:")
        print("- Parameter Name:")
        print("- Number of IF Statement:")
        
        
        if i!=func_cnt-1:
            print("")


def main(argv):
    path = argv[1]
    
    with open(path, 'r', encoding='utf-8') as json_file:
        data = json.load(json_file)
    
    func_cnt = parse_func_cnt(data)
    func_ret = parse_func_ret(data)
    func_name = parse_func_name(data)
    param_type, param_name = parse_param(data)
    if_cnt = parse_if_cnt(data)
    
    print_analysis(func_cnt, func_ret, func_name, param_type, param_name, if_cnt)


if __name__=="__main__":
    argv = sys.argv
    main(argv)

File: analyze_ast.c
#include <stdio.h>
#include "json_c.c"
#include <string.h>

#define MAX 1000000


void parse_func_name(json_value decl){
    char* func_name = json_get_string(json_get(decl, "name"));

    printf("- Name: %s\n", strtok(func_name, "\""));
}


void parse_func_var(json_value decl){
    char* var_name="", var_type="";

    json_value type = json_get(decl, "type");
    json_value args = json_get(type, "args");
    
    if(!json_is_null(args)){
        json_value params = json_get(args, "params");

        printf("- Parameter Name: "); 
        for(int i=0; i<json_len(params); i++){
            char *name = json_get_string(json_get(args, "params", i, "name"));

            printf("%s", strtok(name, "\""));

            if(i!=json_len(params)-1) printf(", ");
            else putchar('\n');
        } 

        printf("- Parameter Type: "); 
        for(int i=0; i<json_len(params); i++){
            json_value param = json_get(args, "params", i, "type");
            json_value type = json_get(param, "type");
            json_value node = json_get(type, "_nodetype");
            
            if (strcmp(json_get_string(node), "TypeDecl")==0){   
                json_value type2 = json_get(type, "type");
                json_value var_type = json_get(type2, "names", 0);
                printf("%s", json_get_string(var_type));
            }else{
                json_value var_type = json_get(type, "names", 0);
                printf("%s", json_get_string(var_type));
            }

            json_value node2 = json_get(param, "_nodetype");
            
            if (strcmp(json_get_string(node2), "PtrDecl")==0) printf(" *");

            if(i!=json_len(params)-1) printf(", ");
            else putchar('\n');
        } 
    }
}


int main(int argc, char* argv[]){
    FILE* fp = fopen(argv[1], "r");  
    char file[MAX] = {0, };
    int func_cnt = 0;

    fread(file, 1, MAX, fp); 
    fclose(fp); 
    
    json_value data = json_create(file);
    json_value ext = json_get(data, "ext");

    for(int i=0; i<json_len(ext); i++){
        json_value node = json_get(data, "ext", i, "_nodetype");
        char* _nodetype = json_get_string(node);

        if (strcmp(_nodetype, "FuncDef")==0){
            func_cnt++;
            printf("Function %d:\n", func_cnt);
            
            json_value body = json_get(data, "ext", i, "body");
            json_value decl = json_get(data, "ext", i, "decl");
            
            parse_func_name(decl);
            parse_func_var(decl);
        }

        putchar('\n');
    }
    
    printf("Total Number of Function: %d\n", func_cnt);

    json_free(data);

    return 0;
}
    
